Parsing Content-Type

annevk opened the pull request:
See https://github.com/whatwg/mimesniff/issues/30#issuecomment-382721437 and https://github.com/whatwg/mimesniff/issues/30#issuecomment-436632541 for the processing model this is meant to test.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

annevk committed:
Parsing Content-Type

Very much WIP.

annevk committed:
add another test

annevk committed:
add script type tests and make server script more reliable

annevk committed:
revert change to check_stability

annevk committed:
test combined and separated

annevk committed:
test some invalid values

annevk committed:
nit

annevk committed:
test empty string too

annevk committed:
escaped quotes

annevk commented:
@MattMenke2 could you review these too?

MattMenke2 commented:
Happy to review the test cases, but I'm not at all familiar with the python+js environment used by the wpt tests, so at least that aspect should probably be reviewed by someone familiar with the test environment.

MattMenke2 commented on fetch/content-type/resources/content-types.json:
```diff
@@ -0,0 +1,104 @@
+    "encoding": null,
+    "mimeType": "text/html",
+    "documentContentType": "text/html"
+  }
```
This seems like a pretty complete list to me.  Some suggestions:

// Make sure values other than charset are discarded from earlier lines (Should we test "boundary", too?  It's the only other value supported in browsers, so could imagine boundary and charset behaving the same, while every other values behaved differently):
"input": ["text/plain;charset=gbk;x=foo", "text/plain"] => "text/plain; charset=GBK"

// Make sure close quotes work across input lines.
"input": ["text/html;\"", "\\\"", "text/plain", "\";charset=GBK"] => "text/html; charset=GBK"

// This is largely the same as above, but for the MIME type itself, and without the extra entire quoted line - don't think it adds a whole lot of value, but still think it's nice to have the close quote and the final value on different lines.  Could also do a variant of this without the "\"", but think the value there is even more minimal than this one.
"input": ["text/html;\"",  "\"", "text/plain"] => "text/plain"

// Make sure that quoted MIME types don't work.  A case testing this perhaps fits better in wpt/mimesniff/mime-types/resources/mime-types.json.
["text/plain", "\"text/html\""] => "text/plain"

MattMenke2 commented on fetch/content-type/resources/script-content-types.json:
```diff
@@ -0,0 +1,87 @@
+    "encoding": null
+  },
+  {
+    "input": ["text / javascript"],
```
Suggest splitting this in two, one with the space in each location.

Also, should the cases be checked (instead or additionally) in wpt/mimesniff/mime-types/resources/mime-types.json?

MattMenke2 commented on fetch/content-type/resources/script-content-types.json:
```diff
@@ -0,0 +1,87 @@
+    "encoding": null
+  },
+  {
+    "input": ["text/javascript;charset=windows-1252", "x/x", "text/javascript"],
```
Should the first one also be of a type that doesn't execute, for this test and the next 3 as well?

annevk replied:
Then the encoding wouldn't copy over.

MattMenke2 commented on fetch/content-type/script.window.js:
```diff
@@ -0,0 +1,48 @@
+    document.head.appendChild(script);
+    if (testData.executes) {
+      script.onload = t.step_func_done(() => {
+        assert_equals(self.isThereABetterWay, testData.encoding === "windows-1252" ? "â‚¬" : "€");
```
Is there a reason for checking the encoding here?  Also, I have no idea what all the isThereABetterWay stuff is doing.

annevk replied:
It's basically making sure the script got decoded correctly. `isThereABetterWay` is a global variable that the script sets with the decoded string value. It's called that way because I was hoping someone would come up with a better mechanism than a global variable. I'll rename it to something more appropriate.

annevk committed:
add more tests

annevk committed:
documentation of JSON resources

annevk commented:
Thanks a lot, these are great. I'm not doing `boundary` for now as that basically requires testing (and defining...) that format, which I want to do at some point, but I'm not gonna block on.

I fixed your other points (or left comments), and added documentation to avoid renaming `input` (and since documentation seems good anyway).

annevk commented:
@domenic note that I added some new generic MIME type tests here as well per feedback, you want to run those against your implementation.

MattMenke2 commented on fetch/content-type/README.md:
```diff
@@ -0,0 +1,20 @@
+
+An array of tests. Each test has these fields:
+
+* `input`: an array of values for the `Content-Type` header. A harness needs to run the test twice if there are multiple values. One time with the values concatenated with `,` followed by a space and one time with multiple `Content-Type` declarations, each on their own line with one of the values, in order.
```
I think think it's better to rename input so the tests are more self-contained.  Note going to block on that, though.

domenic commented:
Updates to mimesniff/mime-types/resources/mime-types.json LGTM. One day maybe we'll have a jsdom implementation of fetch that can fully test the other JSON files.

annevk committed:
input -> contentType

annevk commented:
Fair enough, I renamed `input` to `contentType`.
